server/src/storage/compressed_data_buffer.rs

```diff
@@ -0,0 +1,169 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Buffer for compressed segments from the same table.
+
+use std::fs;
+use std::io::Error as IOError;
+use std::io::ErrorKind::Other;
+use std::path::Path;
+
+use datafusion::arrow::compute;
+use datafusion::arrow::record_batch::RecordBatch;
+
+use crate::storage;
+use crate::storage::StorageEngine;
+use crate::types::CompressedSchema;
+
+/// A single compressed buffer, containing one or more compressed segments and providing
+/// functionality for appending segments and saving all segments to a single Apache Parquet file.
+pub(super) struct CompressedDataBuffer {
+    /// Compressed segments that make up the compressed data in the [`CompressedDataBuffer`].
+    compressed_segments: Vec<RecordBatch>,
+    /// Continuously updated total sum of the size of the compressed segments.
+    pub(super) size_in_bytes: usize,
+}
+
+impl CompressedDataBuffer {
+    pub(super) fn new() -> Self {
+        Self {
+            compressed_segments: Vec::new(),
+            size_in_bytes: 0,
+        }
+    }
+
+    /// Append `compressed_segments` to the [`CompressedDataBuffer`] and return the size of
+    /// `compressed_segments` in bytes. It is assumed that `compressed_segments` is sorted by time.
+    pub(super) fn append_compressed_segments(&mut self, compressed_segments: RecordBatch) -> usize {
+        let segment_size = Self::get_size_of_compressed_segments(&compressed_segments);
+
+        self.compressed_segments.push(compressed_segments);
+        self.size_in_bytes += segment_size;
+
+        segment_size
+    }
+
+    /// If the compressed segments are successfully saved to an Apache Parquet file return [`Ok`],
+    /// otherwise return [`IOError`].
+    pub(super) fn save_to_apache_parquet(
+        &mut self,
+        folder_path: &Path,
+        compressed_schema: &CompressedSchema,
+    ) -> Result<(), IOError> {
+        debug_assert!(
+            !self.compressed_segments.is_empty(),
+            "Cannot save CompressedDataBuffer with no data."
+        );
+
+        // Combine the compressed segments into a single RecordBatch.
+        let batch =
+            compute::concat_batches(&compressed_schema.0, &self.compressed_segments).unwrap();
+
+        // Create the folder structure if it does not already exist.
+        fs::create_dir_all(folder_path)?;
+
+        // Create a new file that includes the start timestamp of the first segment in batch, the
+        // end timestamp of the last segment in batch, the minium value stored in batch, and the
+        // maximum value stored in batch as the file name to efficiently pruning files that only
```

Comment:
maybe `...to support efficiently pruning...` instead.

Reply:
Fixed.